onnx/optimizer/passes/split.h

```diff
@@ -11,9 +11,68 @@ static constexpr const char* impure_operators[] = {
   "RandomNormal",
   "RandomNormalLike",
   "RandomUniform",
-  "RandomUniformLike"
+  "RandomUniformLike",
+  "If",
+  "Loop",
 };
 
+static bool is_safe_to_speculate(Node* n) {
+  return n->kind() == kTranspose;
+}
+
+static void speculate_ops(Graph& graph, std::unordered_map<std::string, Value*>& capture_map) {
+  for(Value* input : graph.inputs()) {
+    capture_map[input->uniqueName()] = input;
+  }
+  for(auto it = graph.nodes().begin(), end = graph.nodes().end();
+      it != end;
+      ++it) {
+    Node * n = *it;
+
+    // prevent overwriting entries in the capture map with the using capture nodes
+    if(n->kind() == Symbol("Captured")) {
+      continue;
+    }
+
+    for(Value* o : n->outputs() ) {
+      capture_map[o->uniqueName()] = o;
+    }
+
+    for(auto s : n->attributeNames()) {
```

Comment:
Can you just use `DescendOnGraphAttributes` here?

https://github.com/onnx/onnx/blob/master/onnx/optimizer/passes/optimize_pass.h#L30

You can capture the capture map in a lambda